src/js/ctrl/export.ts

```diff
@@ -37,29 +37,34 @@ let counts = duplicate(SPEC_COUNT);
  * direct-manipulation interactors (handles, connectors, etc.).
  * @returns {Object} A Vega specification.
  */
-export function exporter(internal: boolean = false): Spec {
-  const state = store.getState();
+export function exporter(internal: boolean = false, visState?: VisStateTree): Spec { // TODO(ej): use history.ts so you only have to pass id or index
+  if (visState && visState.getIn(['marks']).size == 0) {
+    return {};
+  }
+
+  visState = visState || store.getState().getIn(['vis', 'present']);
   const int = internal === true;
 
   counts = duplicate(SPEC_COUNT);
 
-  let spec: Spec = exporter.scene(state, int);
-  spec.data = exporter.pipelines(state, int);
+  let spec: Spec = exporter.scene(visState, int);
+  spec.data = exporter.pipelines(visState, int);
 
   // Add interactions and widgets from store
-  spec = exporter.interactions(state, spec);
-  spec = exporter.widgets(state, spec);
+  spec = exporter.interactions(visState, spec);
+  spec = exporter.widgets(visState, spec);
   return spec;
 }
 
-exporter.interactions = function(state: State, spec) {
-  state.getIn(['vis', 'present', 'interactions']).forEach((interaction: InteractionRecord) => {
-    const group: GroupRecord = state.getIn(['vis', 'present', 'marks', String(interaction.groupId)]);
+exporter.interactions = function(visState: VisStateTree, spec) {
+  let storeState = store.getState(); // use Store state instead
+  visState.getIn(['interactions']).forEach((interaction: InteractionRecord) => {
+    const group: GroupRecord = visState.getIn(['marks', String(interaction.groupId)]);
     const groupName = exportName(group.name);
-    const scaleInfo = getScaleInfoForGroup(state, group._id);
-    const fieldsOfGroup = getFieldsOfGroup(state, group._id);
+    const scaleInfo = getScaleInfoForGroup(storeState, group._id);
+    const fieldsOfGroup = getFieldsOfGroup(storeState, group._id);
```

Comment:
now that i think about it, these should use visState too. if we pass in an older visState to the exporter, we expect it to get the scale info for that state. using the store state will use the current scale info instead, which could cause problems.

this means that the getScaleInfoForGroup and getFieldsOfGroup functions should probably be edited to use visState as well, and all uses of those functions updated accordingly. could you add one more commit with that change?